src/device/mod.rs

```diff
@@ -1316,38 +1315,22 @@ impl Runtime {
     }
 
     async fn connect_exit_node(&mut self, exit_node: &ExitNode) -> Result {
-        let mut node: Node = exit_node.try_into()?;
+        let exit_node = exit_node.clone();
 
         // dns socket for macos should only be bound to tunnel interface when connected to exit,
         // otherwise with no exit dns peer will try to forward packets through tunnel and fail
         bind_tun::set_should_bind(true);
 
-        let wg_peer = self
-            .entities
-            .wireguard_interface
-            .get_interface()
-            .await?
-            .peers
-            .get(&node.public_key)
-            .cloned();
-
-        if let Some(wg_peer) = wg_peer {
-            node.is_vpn = false;
-            node.endpoint = wg_peer.endpoint;
-
+        // if exit_node is not a vpn server
+        if exit_node.endpoint.is_none() {
```

Comment:
I know I mentioned that the the way to decide whether `exit_node` is meshnet peer and we are traffic routing or whether `exit_node` is VPN server is by endpoint being either `None` or `Some(...)` respectively.

But there is alternative way, by checking whether a node with the same public key exists amongst meshnet peers, e.g. node such that `requested_state.meshnet_config.peers.base.public_key == exit_node.public_key`.

We already use similar check in our WG consolidation here: https://github.com/NordSecurity/libtelio/blob/main/src/device/wg_controller.rs#L311 

I am thinking that having single thing being checked would be safer.

Reply:
Hmm it would be also nice to have a check that node that is not in meshnet nodes (assumed to be vpn node).
Has endpoint. nobody curretly forbit's from API point of view, to try to set exit_node, before setting meshnet config.

Reply:
So how about:
```    
    async fn connect_exit_node(&mut self, exit_node: &ExitNode) -> Result {
        let exit_node = exit_node.clone();

        // dns socket for macos should only be bound to tunnel interface when connected to exit,
        // otherwise with no exit dns peer will try to forward packets through tunnel and fail
        bind_tun::set_should_bind(true);

        let mut is_meshnet_exit_node = false;

        if let Some(config) = &self.requested_state.meshnet_config {
            if let Some(peers) = &config.peers {
                is_meshnet_exit_node = peers
                    .iter()
                    .any(|peer| peer.base.public_key == exit_node.public_key)
            }
        }

        if is_meshnet_exit_node {
            if let Some(dns) = &self.entities.dns.lock().await.resolver {
                self.reconfigure_dns_peer(dns, &dns.get_default_dns_servers())
                    .await?;
            }
        } else {
            if exit_node.endpoint.is_none() {
                return Err(Error::EndpointNotProvided);
            }
        }

        self.requested_state.exit_node = Some(exit_node);
        wg_controller::consolidate_wg_state(&self.requested_state, &self.entities).await
    }
```

Reply:
This looks nice to me!